src/datasources/email/email.datasource.ts

```diff
@@ -0,0 +1,157 @@
+    @Inject(IConfigurationService) configurationService: IConfigurationService,
+  ) {
+    this.verificationExpireTimeInSeconds = configurationService.getOrThrow(
+      'email.verificationExpireTimeInSeconds',
+    );
+  }
+
+  private _resetVerificationCode(
+    sql: postgres.Sql,
+    emailId: number,
+  ): postgres.PendingQuery<VerificationStatus[]> {
+    // TODO proper code generation
+    const code = Math.floor(Math.random() * 100_000);
+    return sql<VerificationStatus[]>`
+        INSERT INTO emails.verification (id, verification_code)
+        VALUES (${emailId}, ${code})
+        ON CONFLICT (id) DO UPDATE SET verification_code = ${code},
+                                       sent_on           = now()
+        RETURNING *
+    `;
+  }
+
+  private getEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): postgres.PendingQuery<Email[]> {
+    return this.sql<Email[]>`SELECT *
+                             FROM emails.signer_emails
+                             WHERE chain_id = ${args.chainId}
+                               and safe_address = ${args.safeAddress}
+                               and signer = ${args.signer}`;
+  }
+
+  async saveEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    emailAddress: string;
+    signer: string;
+  }): Promise<EmailVerificationCode> {
+    return await this.sql.begin(async (sql) => {
+      const [email] = await sql<Email[]>`
+          INSERT INTO emails.signer_emails (chain_id, email_address, safe_address, signer)
+          VALUES (${args.chainId}, ${args.emailAddress}, ${args.safeAddress}, ${args.signer})
+          RETURNING *
+      `;
+
+      const [verificationStatus] = await this._resetVerificationCode(
+        sql,
+        email.id,
+      );
+
+      return <EmailVerificationCode>{
+        emailAddress: email.email_address,
+        verificationCode: verificationStatus.verification_code,
+      };
+    });
+  }
+
+  async resetVerificationCode(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+  }): Promise<{ emailAddress: string; verificationCode: string }> {
+    const [email] = await this.getEmail(args);
+
+    if (!email) {
+      throw new EmailAddressDoesNotExistError(
+        args.chainId,
+        args.safeAddress,
+        args.signer,
+      );
+    }
+
+    if (email.verified) {
+      throw new EmailAlreadyVerifiedError(email);
+    }
+
+    const [verificationStatus] = await this._resetVerificationCode(
+      this.sql,
+      email.id,
+    );
+
+    return <EmailVerificationCode>{
+      emailAddress: email.email_address,
+      verificationCode: verificationStatus.verification_code,
+    };
+  }
+
+  async verifyEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    signer: string;
+    code: string;
+  }): Promise<void> {
+    return await this.sql.begin(async (sql) => {
+      const [email] = await this.getEmail(args);
```

Comment:
Same concern as above: this is getting the first of potentially many rows matching the `args`.

Reply:
There's a `UNIQUE` constraint enforced on the database level. We can have a test that validates that the constraint was applied.

Are you suggesting validating the collection size in addition to that?

Reply:
Same as in the above comment, the email address seems to be part of the unique constraint:
```
UNIQUE (chain_id, email_address, safe_address, signer)
```

Maybe I'm missing something 🤔